image: Refactor to use cas/ref engines instead of walkers

wking opened the pull request:
Abstract out the details of the image-layout format, since the
validators and unpackers shouldn't care about that sort of backend
stuff.  This should make it easy to build backends based on zip files,
HTTP, the Docker registry format, etc.  See #140.

I ran out of time on the final commit, but wanted to push this in case
folks had time for early feedback.  I should have the last commit
fixed up today, but there may be some more rerolling while I do that.

Ping @s-urbaniak.

wking commented:
I haven't tested all the old commands yet, but 9252dee → aba65b0 fixes the obvious-to-me-from-reading-the-code issues, so I've removed the WIP label.

wking commented:
I'd forgotten to remove `image/walkers.go` now that we don't need it anymore.  Done with aba65b0 → b4482da.

s-urbaniak commented on cmd/oci-image-tool/cas_get.go:
```diff
@@ -0,0 +1,94 @@
+	}
+}
+
+func (state *casGetCmd) Run(cmd *cobra.Command, args []string) {
```
just call this `cmd`

wking replied:
On Mon, Jun 20, 2016 at 05:30:34AM -0700, Sergiusz Urbaniak wrote:
> > +func (state *casGetCmd) Run(cmd *cobra.Command, args []string) {
> 
> just call this `cmd`

There are a few things in that line.  Do you mean ‘Run’ → ‘cmd’?

s-urbaniak replied:
Sorry for being inprecise. I meant the `state` variable, I'd call this `cmd`.

wking replied:
On Mon, Jun 20, 2016 at 02:09:20PM -0700, Sergiusz Urbaniak wrote:
> > +func (state *casGetCmd) Run(cmd *cobra.Command, args []string) {
> 
> Sorry for being inprecise. I meant the `state` variable, I'd call this `cmd`.

What do you want me to use for *cobra.Command?

s-urbaniak replied:
Haha, sorry, it's 11pm here in old Europe, my eyes are starting to blur ;-) In this case I'd recommend calling it `getCmd`, or even `cgc` which is quite idiomatic.

wking replied:
On Mon, Jun 20, 2016 at 02:19:47PM -0700, Sergiusz Urbaniak wrote:
> > +func (state *casGetCmd) Run(cmd *cobra.Command, args []string) {
> 
> Haha, sorry, it's 11pm here in old Europe, my eyes are starting to
> blur ;-) In this case I'd recommend calling it `getCmd`, or even
> `cgc` which is quite idiomatic.

getCmd would cause collisions between the current casGetCmd and
refsGetCmd.  Do you really prefer types ‘cgc’ and ‘rgc’ to casGetCmt
and refsGetCmd?  I can reroll (it's not my project ;), but the short
versions are a bit dense for my taste ;).

s-urbaniak commented on cmd/oci-image-tool/refs_get.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func (state *refsGetCmd) run() (err error) {
+	engine, err := layout.GetEngine(state.path)
```
pleas, no getters, see https://golang.org/doc/effective_go.html#Getters. Since you commented that this method "guesses", just call it `GuessEngine`.

runcom replied:
you need to `defer engine.Close()` here, don't you?

wking replied:
On Mon, Jun 20, 2016 at 05:32:17AM -0700, Sergiusz Urbaniak wrote:
> > +	engine, err := layout.GetEngine(state.path)
>
> pleas, no getters, see
> https://golang.org/doc/effective_go.html#Getters. Since you
> commented that this method "guesses", just call it `GuessEngine`.

It's not really a getter (in the method to return an instance
attribute sense).  Maybe NewEngine would be more idiomatic for an
instantiator?

runcom replied:
+1 on NewEngine (even if I like GuessEngine)

s-urbaniak commented on cmd/oci-image-tool/refs_get.go:
```diff
@@ -0,0 +1,81 @@
+		return err
+	}
+
+	descriptor, err := engine.Get(state.name)
```
You are returning a `Descriptor`, hence I'd call this method simply `engine.Descriptor(cmd.name)`

s-urbaniak replied:
Nevermind, in this case I see you're consistent with the names across engines, here it makes sense.

s-urbaniak commented on cmd/oci-image-tool/refs_get.go:
```diff
@@ -0,0 +1,81 @@
+	}
+
+	encoder := json.NewEncoder(state.stdout)
+	return encoder.Encode(&descriptor)
```
just flatten both lines.

s-urbaniak commented on image/cas/interface.go:
```diff
@@ -0,0 +1,37 @@
+type Engine interface {
+
+	// Put adds a new blob to the store.
+	Put(writer io.Writer) (digest string, err error)
```
I guess `Put` needs to take a `io.Reader` where the engine reads the blob from rather than a writer?

s-urbaniak commented on image/cas/layout/interface.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+// ReadSeekCloser wraps the Read, Seek, and Close methods.
+type ReadSeekCloser interface {
```
```
type ReadSeekCloser interface {
	io.ReadSeeker
	io.Closer
}
```

s-urbaniak commented on image/cas/layout/main.go:
```diff
@@ -0,0 +1,36 @@
+		return nil, err
+	}
+
+	return GetTarEngine(file)
```
Please no getters, I'd call this `newTarEngine`.

s-urbaniak commented on image/refs/layout/main.go:
```diff
@@ -0,0 +1,36 @@
+
+// GetEngine guesses the appropriate backend (tar, HTTP, ...) and
+// returns that engine.
+func GetEngine(path string) (engine refs.Engine, err error) {
```
`GuessEngine`

s-urbaniak commented on image/cas/layout/main.go:
```diff
@@ -0,0 +1,36 @@
+
+// GetEngine guesses the appropriate backend (tar, HTTP, ...) and
+// returns that engine.
+func GetEngine(path string) (engine cas.Engine, err error) {
```
`GuessEngine`

s-urbaniak commented on image/structure/image_layout.go:
```diff
@@ -0,0 +1,21 @@
+
+// ImageLayoutVersion represents the oci-version content for the image
+// layout format.
+type ImageLayoutVersion struct {
```
Do we want these types really exported?

wking replied:
On Mon, Jun 20, 2016 at 05:51:24AM -0700, Sergiusz Urbaniak wrote:
> > +// ImageLayoutVersion represents the oci-version content for the image
> > +// layout format.
> > +type ImageLayoutVersion struct {
> 
> Do we want these types really exported?

This is part of the spec [1], so I think we do want to export it.  We
probably also want a JSON Schema for that content as well, in which
case the public type will probably end up being generated by #147 or
similar.

[1]: https://github.com/opencontainers/image-spec/blame/c3ad3c723522b1155678b5c5c83971fc2a98d7ca/image-layout.md#L15

s-urbaniak commented on image/structure/doc.go:
```diff
@@ -0,0 +1,16 @@
+// limitations under the License.
+
+// Package structure defines stucts for image-spec content.
+package structure
```
There's merely on struct defined here. Since we wanted the oci-image repo to include only schemas, I'd be careful exposing any structs.

wking replied:
On Mon, Jun 20, 2016 at 06:01:08AM -0700, Sergiusz Urbaniak wrote:
> > +// Package structure defines stucts for image-spec content.
> > +package structure
> 
> There's merely on struct defined here. Since we wanted the oci-image
> repo to include only schemas, I'd be careful exposing any structs.

Two structures: Descriptor and ImageLayoutVersion.  And #147 is
planning on exporting auto-generated versions, in which case we can
drop these hand-coded structures and adjust the cas/ref engines to use
those.  Is “not exposing structures for handling JSON defined in this
spec” on the table?  I think #147 is “should we autogenerate these or
write them by hand?” and not “should we expose these at all?”.

runcom commented on image/image.go:
```diff
@@ -19,136 +19,132 @@ import (
+// Validate validates the given reference.
+func Validate(path, ref string) error {
+	refEngine, err := refsLayout.GetEngine(path)
+	defer refEngine.Close()
```
defer should be called after err checking

runcom commented on image/image.go:
```diff
@@ -19,136 +19,132 @@ import (
-func Validate(tarFile, ref string) error {
-	f, err := os.Open(tarFile)
+	casEngine, err := casLayout.GetEngine(path)
+	defer casEngine.Close()
```
ditto (defer after err check)

runcom commented on image/image.go:
```diff
@@ -19,136 +19,132 @@ import (
+// Unpack unpacks the given reference to a destination directory.
+func Unpack(path, dest, ref string) error {
+	refEngine, err := refsLayout.GetEngine(path)
+	defer refEngine.Close()
```
defer close after err check

runcom commented on image/image.go:
```diff
@@ -19,136 +19,132 @@ import (
-func Unpack(tarFile, dest, ref string) error {
-	f, err := os.Open(tarFile)
+	casEngine, err := casLayout.GetEngine(path)
+	defer casEngine.Close()
```
defer close after err check

runcom commented on image/image.go:
```diff
@@ -19,136 +19,132 @@ import (
+// destination.
+func CreateRuntimeBundle(path, dest, ref, rootfs string) error {
+	refEngine, err := refsLayout.GetEngine(path)
+	defer refEngine.Close()
```
there are a bunch of these

runcom commented on cmd/oci-image-tool/cas_get.go:
```diff
@@ -0,0 +1,94 @@
+		os.Exit(1)
+	}
+
+	os.Exit(1)
```
shouldn't be this `os.Exit(0)` at this point?

runcom commented on cmd/oci-image-tool/refs_get.go:
```diff
@@ -0,0 +1,81 @@
+		os.Exit(1)
+	}
+
+	os.Exit(1)
```
`os.Exit(0)`?

runcom commented on cmd/oci-image-tool/refs_list.go:
```diff
@@ -0,0 +1,88 @@
+		os.Exit(1)
+	}
+
+	os.Exit(1)
```
`os.Exit(0)`?

runcom commented on image/structure/descriptor.go:
```diff
@@ -0,0 +1,22 @@
+package structure
+
+// Descriptor represents application/vnd.oci.descriptor.v1.
+type Descriptor struct {
```
I hope we'll be moving this out to generated types soon

runcom commented on cmd/oci-image-tool/refs_list.go:
```diff
@@ -0,0 +1,88 @@
+	if err != nil {
+		return err
+	}
+
```
`defer engine.Close()`?

runcom commented on image/image.go:
```diff
@@ -19,136 +19,132 @@ import (
-func CreateRuntimeBundle(tarFile, dest, ref, root string) error {
-	f, err := os.Open(tarFile)
+	casEngine, err := casLayout.GetEngine(path)
+	defer casEngine.Close()
```
defer after err check

wking commented:
I just pushed 19697a2 → 1c276b6 addressing most of the first round's feedback:

* [Flatten `NewEncoder`/`Encode`][1].
* Fix [`Writer` → `Reader` for `cas.Engine.Put`][2].
* [Use `ReadSeeker` in `ReadSeekCloser`][3].
* [`GetTarEngine` → `NewTarEngine`][4] (@s-urbaniak suggested
  `newTarEngine`, but I think we might as well expose it for
  folks who *know* they need a tar-backed version of the
  image-layout engine and don't want to bother with
  auto-detection).
* [`GetEngine` → `NewEngine`][8].
* [`defer` after error checking][5].
* [Final `Exit(1)` → `Exit(0)` in new oci-image-tool commands][6].
* [`defer engine.Close()` in new oci-image-tool commands][7].

The outstanding issues from the first round of review are:

* [Rename something to `cmd`][9].  This is just blocked on me understanding which token gets renamed.
* Talking about what structures to export ([here][10] and [here][11]).  I've left my thought in responses there.

[1]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67681624
[2]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67681864
[3]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67682062
[4]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67682110
[5]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67684851
[6]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67685905
[7]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67686714
[8]: https://github.com/opencontainers/image-spec/pull/159/files/19697a23ba6ce4116349e83ce1733e8e0a2880d9#r67762200
[9]: https://github.com/opencontainers/image-spec/pull/159/files#r67681179
[10]: https://github.com/opencontainers/image-spec/pull/159/files#r67684770
[11]: https://github.com/opencontainers/image-spec/pull/159/files#r67683568

wking commented:
Rebased onto the current master with 1c276b6 → 2e09496 to pick up #148, avoiding unrelated broken tests due to #162.

wking commented:
Remaining Travis error is [a lint duplication complaint][1].  If anyone has clever ideas for deduping my new `cmd/oci-image-tool/…` files, I'm happy to adjust them.

[1]: https://travis-ci.org/opencontainers/image-spec/builds/140542858#L234

wking commented:
I pushed 2e09496 → 077bb4b which:

* Renamed `CheckVersion` to the more specific `CheckTarVersion` (making room for zip-backed version checkers, directory-backed version checkers, etc.).
* Replaced a `i += 1` with `i++`, as requested by golint.
* Told our lint script to not worry about the [duplication complaints][1].

[1]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-228757555

sgotti commented on image/cas/interface.go:
```diff
@@ -0,0 +1,37 @@
+	Put(writer io.Reader) (digest string, err error)
+
+	// Get returns a reader for retrieving a blob from the store.
+	Get(digest string) (reader io.Reader, err error)
```
Sometimes an io.ReadCloser is needed. Suppose an engine that writes/reads every blob to a file, the caller should close the io.ReaderCloser.

wking replied:
Ah, right.  Will update this in my next reroll.

sgotti commented on image/refs/interface.go:
```diff
@@ -0,0 +1,65 @@
+	Put(name string, descriptor *structure.Descriptor) (err error)
+
+	// Get returns a reference from the store.
+	Get(name string) (descriptor *structure.Descriptor, err error)
```
If the reference doesn't exists, descriptor will be nil or an error will be returned?

wking replied:
On Fri, Jul 01, 2016 at 09:04:29AM -0700, Simone Gotti wrote:
> > +	// Get returns a reference from the store.
> > +	Get(name string) (descriptor *structure.Descriptor, err error)
> 
> If the reference doesn't exists, descriptor will be nil or an error
> will be returned?

Ah, good point.  Currently this returns io.EOF as an error [1,2]
because that's what tar does [3].  But that's probably not an
intuitive error for a more generic “key not found” problem.  Is there
a stock Go error that's a better match, or should we create our own?
If we create our own, do we want separate error variables for the cas
and refs interfaces?

And however this goes down, we'll want to document the key-not-found
case here.

[1]: https://github.com/opencontainers/image-spec/pull/159/files#diff-b11cc21c1cf82254929b0e2e98cbd695R74
[2]: https://github.com/opencontainers/image-spec/pull/159/files#diff-ce780f6bc5726d182fefbd55676f47eaR70
[3]: https://golang.org/pkg/archive/tar/#Reader.Next

sgotti replied:
Go os pkg returns an error on a file not found and provides [os.IsNotExists(err)](https://golang.org/pkg/os/#IsNotExist) the check it.

It depends if the not found should be considered an error or not.

I usually like to return a nil entry if I want to handle the not found case without relying on an error.

But in this case refsEnigine.Delete will return an error on missing name, so probably an IsNotExists(err) functions or a ErrNotFound should be defined anyway. 

And how the same case should be implemented in the casengine.Get function?

wking commented:
I pushed 077bb4b → 53c876b which:

* Fixed a duplicate `engine.Close` in `cas_get.go`.
* Documented the current [`io.EOF` missing-key errors][1], although I'm fine [returning a different error][2] if folks want something different.
* [Return an `io.ReadCloser`][3] from `cas.Engine.Get`.

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r69321782
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r69333604
[3]: https://github.com/opencontainers/image-spec/pull/159#discussion_r69316937

wking commented on image/refs/interface.go:
```diff
@@ -0,0 +1,66 @@
+	// * List("", 2, 0) -> ["123", "abcd"]
+	// * List("", 2, 1) -> ["abcd", "abce"]
+	// * List("abc", 2, 1) -> ["abce", "abcf"].
+	List(prefix string, size int, from int) (names []string, err error)
```
There might be better API here (for large ref sets) if we return a channel, so folks can launch a `List` goroutine and iterate over names without having to construct temporary slices.  That opens the door to iterating over all names without paginating, because you wouldn't have to hold a huge `names` slice in memory at once.

I think we want to keep the pagination arguments though, so callers can efficiently jump to a particular position or launch several parallel `List` calls to walk the namespace (assuming the backend efficiently implements the pagination arguments).

wking commented:
I pushed 53c876b → 0f46052 which:

* Converts the missing-key error from `io.EOF` [to `os.ErrNotExist`][1].
* Adds a new commit converting `refs.Engine.List` to [use goroutines and channels][2].

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r69388563
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r69380750

sgotti commented on image/cas/interface.go:
```diff
@@ -0,0 +1,39 @@
+type Engine interface {
+
+	// Put adds a new blob to the store.
+	Put(writer io.Reader) (digest string, err error)
```
Since there isn't mediaType information provided in put and returned by get, should the user autodetect the mediaType or is this just not needed since it'll be already known because the user will always fetch a blob starting from a parent referencing object (descriptor, manifest, manifest list) that already provides the mediaType?

This assumes that all the objects will be part of a tree and there's always the way to retrieve the mediaType by the parent.

Instead if there isn't this assumption then a mediaType should be returned or it should be detected by the user for every object. (another example can be a List method that can return blobs by mediaTypes)

wking replied:
On Mon, Jul 04, 2016 at 01:44:34AM -0700, Simone Gotti wrote:
> > +	// Put adds a new blob to the store.
> > +	Put(writer io.Reader) (digest string, err error)
> 
> Since there isn't mediaType information provided in put and returned
> by get, should the user autodetect the mediaType or is this just not
> needed since it'll be already known because the user will always
> fetch a blob starting from a parent referencing object (descriptor,
> manifest, manifest list) that already provides the mediaType?

I don't think it's needed for consuming code.  But we might need this
information for a Docker-registry-based backend, because where you
push the data depends on what you're pushing [1].

> This assumes that all the objects will be part of a tree and there's
> always the way to retrieve the mediaType by the parent.

It doesn't necessarily mean all blobs will be part of a tree, it just
assumes that if you know you want blob $DIGEST, you already know it's
type.  In a world where all Merkle links are descriptors [2], that's a
safe assumption.

> Instead if there isn't this assumption then a mediaType should be
> returned or it should be detected by the user for every
> object. (another example can be a List method that can return blobs
> by mediaTypes)

I don't think list-CAS-blobs-by-media-type is an API we need to
support.  In oci-store, you only use ListBlobs (which filters by
media-type) for garbage collection, and you don't need that if you
drill down from refs.

[1]: https://docs.docker.com/registry/spec/api/#/pushing-an-image
     Layers to blobs/uploads/, manifests to manifests/<reference>, …
[2]: https://github.com/opencontainers/image-spec/blame/ca58cdfc7c8200c6c62287e677fadb0074cc29cc/descriptor.md#L11

stevvooe replied:
The interface doesn't need to be nearly as heavy weight, but I'd recommend taking inspiration from the [distribution interfaces](https://godoc.org/github.com/docker/distribution#BlobWriter). They support a number of uses that aren't obvious and there are few mistakes made there, as well:

1. It's much easier to control the copy if you have something to write to, rather than passing in a writer and hoping for the best.
2. Consider the case of re-entrant writes. I don't think they are necessary here, but having a repeatable identifier for resuming a failed write exists there.
3. Consider the transaction boundary at the point this write completes. This is interesting when two writers are trying to write the same blob. This must ensure that only one succeeds but they may race each other.
4. Consider the use of handles, rather than directly accessing the readers and writers. This will allow you to have better control over the scope of a reader and writer. 
5. Be very careful with `ReadFrom`, if you rely on it for performance. The last point helps here but this becomes a nightmare if you wrap interfaces.
6. Consider splitting the front end interface from the driver. Make the front end concrete. There are many details that are important for the driver that aren't important for the user of the interface. (We made this mistake in distribution).

All in all, this looks fine, but I hope these points can help.

Camlistore also is a great resource for structuring these kinds of interfaces, as well.

wking replied:
On Thu, Jul 07, 2016 at 02:53:36PM -0700, Stephen Day wrote:
> 1. It's much easier to control the copy if you have something to
>    write to, rather than passing in a writer and hoping for the best.
> 2. Consider the case of re-entrant writes. I don't think they are
>    necessary here, but having a repeatable identifier for resuming a
>    failed write exists there.

Yeah, I think that's more than we want to bite off here.  And I'm not
sure how much of that is useful to expose to the generic caller.  For
example, you could handle resumable uploads *inside* the engine.
You'd need to hash the content internally to know which ID to pick
from your cache of previously-started-and-failed pushes, but I'm not
sure how you avoid that by exposing the create/resume information to
the caller.  I'll dig into the docker/distribution code and see where
it's getting the resume IDs from to confirm this approach.

> 3. Consider the transaction boundary at the point this write
>    completes. This is interesting when two writers are trying to
>    write the same blob. This must ensure that only one succeeds but
>    they may race each other.

Why does only one have to succeed?  If the blob being put is now in
the store, that sounds like success to me (so idempotent Put).  It's
probably worth documenting that so callers are all on the same page;
I'll push an update shortly.

> 4. Consider the use of handles, rather than directly accessing the
>    readers and writers. This will allow you to have better control
>    over the scope of a reader and writer.
> 5. Be very careful with `ReadFrom`, if you rely on it for
>    performance…

These sound like details for production engine implementations, and I
think the image-spec implementations should try and be simple and
robust reference implementations (even if that means they are slow and
inefficient).  Folks who want a production engine should be writing
that in a separate repository, and they can put their wrapper around
the Reader/Writer inside the engine without changing the Engine
interface.

> 6. Consider splitting the front end interface from the driver. Make
>    the front end concrete. There are many details that are important
>    for the driver that aren't important for the user of the
>    interface. (We made this mistake in distribution).

I'm not sure what you mean by “make the frontend concrete”, but the
goal of this PR is to hide all those messy details behind a generic
interface.

> Camlistore also is a great resource for structuring these kinds of
> interfaces, as well.

The resumable Camlistore interface seems to be here [1], and uses an
opaque continuation token.  I'm not clear on how you recover tokens
after failed pushes, but will look more later.

Both the docker/distribution and Camilstore interfaces take Context
arguments [2] for easy canceling.  I'll mimic that here (and replace
the ‘done’ from refs.Engine.List).

[1]: https://godoc.org/camlistore.org/pkg/blobserver#BlobStreamer
[2]: https://godoc.org/golang.org/x/net/context#Context

stevvooe commented on image/cas/layout/tar.go:
```diff
@@ -0,0 +1,95 @@
+
+	tarReader := tar.NewReader(engine.reader)
+	for {
+		header, err := tarReader.Next()
```
Rather than doing an `O(N)` search through the tar every time, it is sufficient to make a single pass over the tar, noting the offset and length of the target file. When searching for the file in Get, this becomes and index lookup and seek, rather than having to scan the whole file.

This obviously becomes more tricky with compression, so temporary files may be a better approach.

@vbatts may have more detailed information on doing this correctly.

wking replied:
On Thu, Jul 07, 2016 at 02:57:35PM -0700, Stephen Day wrote:
> > +	tarReader := tar.NewReader(engine.reader)
> > +	for {
> > +		header, err := tarReader.Next()
>
> Rather than doing an `O(N)` search through the tar every time, it is
> sufficient to make a single pass over the tar, noting the offset and
> length of the target file. When searching for the file in Get, this
> becomes and index lookup and seek, rather than having to scan the
> whole file.

I'm not really worried about performance here [1].  Iterating each
time is the simplest approach that works, so I think it's a good match
for the spec repo's reference implementation.  Folks who intend to use
the Engine interfaces in production will probably want to pick up an
external implementation aimed at performance.  If I was using this in
production I would certainly not be using an implementation based on
tar ;).

That said, I'm happy to put in straightforward optimizations if they
wouldn't distract from the reference-implementation goal.

[1]: https://github.com/opencontainers/image-spec/pull/159/files#r70000054

wking commented:
I pushed fc8bcc1 → 8a958c9 which:

* Documents the idempotent Put() [1].
* Adds Context for canceling methods [1].

Still open from the most recent rounds of review:

* Whether cas.Engine should know about media types [2].
* Whether resumable push/pull can be handled inside a cas.Engine with
  the current API [1].

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r70000054
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r69426044

wking commented:
On Fri, Jul 08, 2016 at 01:23:15AM -0700, W. Trevor King wrote:
> * Whether resumable push/pull can be handled inside a cas.Engine
>   with the current API [1].

It looks like docker/distribution's reference client does not handle
this [1].

I can't find any resume functionality in docker/docker either
(although I'm not familar with the codebase, maybe I'm just missing
it?).  I did find what seems to be the stock push implementation, and
it's hashing the blobs internally *while* pushing (which is more
efficient) but not using resume [2].

I was also unable to turn up an example of persisting the continuation
token by grepping for StreamBlobs in camlistore/camlistore.

So is there anyone actually using this feature that I can look at?
Can someone more familiar with any of the other projects point me at
the appropriate code?

[1]: https://github.com/docker/distribution/blob/v2.5.0-rc.2/registry/client/repository.go#L760-L762
[2]: https://github.com/docker/docker/blob/v1.12.0-rc3/plugin/distribution/push.go#L67-L95

wking commented:
Rebased around #172 (dropping my Descriptor export from 3e8ea16) with
d1efc62 → 380ee1d.

philips commented on image/cas/interface.go:
```diff
@@ -0,0 +1,43 @@
+// limitations under the License.
+
+// Package cas implements generic content-addressable storage.
+package cas
```
@wking Can you add a note that says "This CAS interface is for internal use of oci-image-tool for the time being. It is subject to change. This notice will be removed when and if the interface becomes stable."

wking replied:
On Wed, Jul 27, 2016 at 02:16:23PM -0700, Brandon Philips wrote:
> @wking Can you add a note that says "This CAS interface is for
> internal use of oci-image-tool for the time being. It is subject to
> change. This notice will be removed when and if the interface
> becomes stable."

Done in 059d807.

wking commented:
I just pushed 9f8a18e → 059d807 which:

* Added unsable caveats to the Engine interfaces [1].
* Rebased onto the current master.
* Vendored golang.org/x/net/context (f3c0dc3)
* Added a longer description to the ‘cas get’ command.
* Transitioned the tar-based engines to use a ReadWriteSeekCloser
  .file instead of a ReadSeekCloser .reader, so we can expand to
  implement writes without too much churn.

[1]: https://github.com/opencontainers/image-spec/pull/159/files/9f8a18eb1cb2a1cbccd6fe137d743dccc2142901#r72525379

wking commented:
I just pushed 059d807 → 2d33ad9 which:

* Fixes a few ‘go fmt’ errors.
* Implements Engine.Put for the tar-backed image-layout CAS
  implementation.  There's some discussion of the implementation in
  the 2d33ad9 commit message.

With this change, the only missing piece of the tar-backed
image-layout CAS implementation is Delete.

wking commented:
I just pushed 2d33ad9 → 61c4bf0 which:

* Adds Typeflag to files created by writeTarEntry.
* Implements refs.Engine.Put (4ed5f51).
* Adds an image-layout creation helper (61c4bf0).

Still no Engine.Delete support.  Maybe I'll get to that later tonight.

wking commented:
I just pushed 61c4bf0 → 528616a which moves the dupl update [1] later
in the stack and adds ignore rules for the put commands.

[1]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-229270533

wking commented:
I just pushed 528616a → 0dc9635 to address [1]:

  image/layout/tar.go:94::warning: declaration of err shadows declaration at tar.go:76: (vetshadow)
  image/layout/tar.go:140::warning: declaration of err shadows declaration at tar.go:138: (vetshadow)

[1]: https://travis-ci.org/opencontainers/image-spec/builds/148234574#L248

wking commented:
Rebased around #177 and #192 with 0dc9635 → b2f7b3b.

wking commented:
Actually fixed the err shadowing with b2f7b3b → 1976fd6, and Travis is
happy [1].

[1]: https://travis-ci.org/opencontainers/image-spec/builds/151358712

wking commented:
Except for [naming][1], I'm not aware of any outstanding issues with this PR.

Would adding Go and/or command line tests make this easier to review (because it would demonstrate correct behavior) or harder to review (because it's more code to look at)?  If you want command-line tests, do you have a harness preference (bats, sharness, …)?  If you'd rather not have tests in this PR, I'm happy to file them in a follow-up PR.  Anyhow, lots of choices for tests, and I'm fine with all of them ;).  Just let me know which way to go.

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r67681179

xiekeyang commented:
@opencontainers/image-spec-maintainers, what's the PR left to do? I'm reading the PR and think it is great.

stevvooe commented on cmd/oci-image-tool/autodetect.go:
```diff
@@ -43,7 +42,7 @@ func autodetect(path string) (string, error) {
 	}
 
 	if fi.IsDir() {
-		return typeImageLayout, nil
```
This doesn't seem right to just remove. Was this not correct before?

wking replied:
On Mon, Aug 29, 2016 at 06:52:32PM -0700, Stephen Day wrote:
> > @@ -43,7 +42,7 @@ func autodetect(path string) (string, error) {
> >  	}
> >  
> >  	if fi.IsDir() {
> > -		return typeImageLayout, nil
> 
> This doesn't seem right to just remove. Was this not correct before?

The new code handles the distinction between typeImageLayout and
typeImage (which is whether the target is a directory or tarball)
inside the NewEngine constructors [1,2].  I haven't added support for
the directory-based engines yet, since there was already enough going
on.  But I'm happy to add them if it would help with review.

[1]: https://github.com/opencontainers/image-spec/pull/159/files/1976fd6aaf0ef41527bf0434e7644310cc553a7d#diff-c3a5563e81df7aa2a94d70642ba7c3ceR28
[2]: https://github.com/opencontainers/image-spec/pull/159/files/1976fd6aaf0ef41527bf0434e7644310cc553a7d#diff-65bd80b0d4865c33446593b6c8e353f3R28

stevvooe commented on image/cas/interface.go:
```diff
@@ -0,0 +1,47 @@
+	// Put adds a new blob to the store.  The action is idempotent; a
+	// nil return means "that content is stored at DIGEST" without
+	// implying "because of your Put()".
+	Put(ctx context.Context, writer io.Reader) (digest string, err error)
```
`writer`? This naming looks backwards.

stevvooe commented on image/cas/interface.go:
```diff
@@ -0,0 +1,47 @@
+// This interface is for internal use of oci-image-tool for the time
+// being.  It is subject to change.  This notice will be removed when
+// and if the interface becomes stable.
+type Engine interface {
```
If this interface is useful, I'd expect to see some generic algorithms in here. For example:

```
Copy(dstEngine, srcEngine Engine, digest string) error
```

wking replied:
On Mon, Aug 29, 2016 at 06:57:28PM -0700, Stephen Day wrote:
> > +// Engine represents a content-addressable storage engine.
> > +//
> > +// This interface is for internal use of oci-image-tool for the time
> > +// being.  It is subject to change.  This notice will be removed when
> > +// and if the interface becomes stable.
> > +type Engine interface {
> 
> If this interface is useful, I'd expect to see some generic
> algorithms in here. For example:
> 
> ```
> Copy(dstEngine, srcEngine Engine, digest string) error
> ```

That doesn't seem like an behavior that needs to be implemented by
each Engine implementation.  Instead, I'd rather add a helper that
implemented Copy on top of the generic Engine interface (and thus
worked with all implementations).  I already do this with
image/cas/put.go.  Should I add an image/cas/copy.go with your copy as
well?

stevvooe replied:
That is not at all what I said. I am not suggesting a new method. `Copy` would be a generic implementation over the `Engine` interface.

wking replied:
On Tue, Aug 30, 2016 at 01:38:05PM -0700, Stephen Day wrote:
> That is not at all what I said. I am not suggesting a new
> method. `Copy` would be a generic implementation over the `Engine`
> interface.

Ok, so exactly the pattern I already use in image/cas/put.go for
PutJSON.  Would you like me to add image/cas/copy.go with your:

  Copy(dstEngine, srcEngine Engine, digest string) error

?

stevvooe replied:
`PutJSON` is an example of a generic util function I expected. If you have use for `Copy`, go ahead and implement it.

wking replied:
On Tue, Aug 30, 2016 at 02:23:30PM -0700, Stephen Day wrote:
> `PutJSON` is an example of a generic util function I expected. If
> you have use for `Copy`, go ahead and implement it.

I don't have a use for Copy yet.  If you're happy with PutJSON, I'll
just leave that as it stands.

stevvooe commented on image/refs/interface.go:
```diff
@@ -0,0 +1,76 @@
+	// * List(ctx, "", 2, 0) -> "123", "abcd"
+	// * List(ctx, "", 2, 1) -> "abcd", "abce"
+	// * List(ctx,"abc", 2, 1) -> "abce", "abcf"
+	List(ctx context.Context, prefix string, size int, from int) (names <-chan string, errc <-chan error)
```
Don't use channels for this.

wking replied:
On Mon, Aug 29, 2016 at 06:58:41PM -0700, Stephen Day wrote:
> > +	//
> > +	// For example, a store with names like:
> > +	//
> > +	// * 123
> > +	// * abcd
> > +	// * abce
> > +	// * abcf
> > +	// * abcg
> > +	//
> > +	// will have the following call/result pairs:
> > +	//
> > +	// * List(ctx, "", -1, 0) -> "123", "abcd", "abce", "abcf", "abcg"
> > +	// * List(ctx, "", 2, 0) -> "123", "abcd"
> > +	// * List(ctx, "", 2, 1) -> "abcd", "abce"
> > +	// * List(ctx,"abc", 2, 1) -> "abce", "abcf"
> > +	List(ctx context.Context, prefix string, size int, from int) (names <-chan string, errc <-chan error)
> 
> Don't use channels for this.

Motivation for channels is in the 1b1f4458 commit message.  Can you
poke some holes in that motivation and/or suggest an alternative?

stevvooe replied:
You're trading off allocation cycles for something that requires lock without even benchmarking to know what is faster. If you don't want allocations, use the call back model.

Pipeline, per that block post, is for adding concurrency to solve a problem. This doesn't need concurrency.

wking replied:
On Tue, Aug 30, 2016 at 01:40:36PM -0700, Stephen Day wrote:
> > +	List(ctx context.Context, prefix string, size int, from int) (names <-chan string, errc <-chan error)
>
> You're trading off allocation cycles for something that requires
> lock without even benchmarking to know what is faster. If you don't
> want allocations, use the call back model.

Callbacks work for me; will reroll tonight.

stevvooe commented on image/refs/interface.go:
```diff
@@ -0,0 +1,76 @@
+	// Put adds a new reference to the store.  The action is idempotent;
+	// a nil return means "that descriptor is stored at NAME" without
+	// implying "because of your Put()".
+	Put(ctx context.Context, name string, descriptor *specs.Descriptor) (err error)
```
Don't pass the `Descriptor` type as a pointer type. Treat as a value.

wking replied:
On Mon, Aug 29, 2016 at 06:59:21PM -0700, Stephen Day wrote:
> > +	// Put adds a new reference to the store.  The action is idempotent;
> > +	// a nil return means "that descriptor is stored at NAME" without
> > +	// implying "because of your Put()".
> > +	Put(ctx context.Context, name string, descriptor *specs.Descriptor) (err error)
> 
> Don't pass the `Descriptor` type as a pointer type. Treat as a value.

Why?  It seems like we might as well save a copy operation, although
Descriptor isn't very big.  Is there concern about error-checking nil
values?  Or that Put somehow mutates the passed-in descriptor?  Or
some other reason?

stevvooe replied:
Do you argue with every single point made in a code review?

wking replied:
On Tue, Aug 30, 2016 at 01:41:16PM -0700, Stephen Day wrote:
> > +	// Put adds a new reference to the store.  The action is idempotent;
> > +	// a nil return means "that descriptor is stored at NAME" without
> > +	// implying "because of your Put()".
> > +	Put(ctx context.Context, name string, descriptor *specs.Descriptor) (err error)
> 
> Do you argue with every single point made in a code review?

Nope, I picked up the ones that were clear without discussion [1].  I
just ask for motivation where the suggestion doesn't match my
preferred behavior.  In this case, I explained my reason for
preferring a pointer (save a copy [2]), and asked for your reason for
preferring pass-by-value.

[1]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243330295
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76726957

stevvooe commented on image/manifest.go:
```diff
@@ -27,100 +27,72 @@ import (
 type manifest struct {
-	Config descriptor   `json:"config"`
-	Layers []descriptor `json:"layers"`
+	Config *specs.Descriptor  `json:"config"`
```
No need for ptr here.

wking replied:
On Mon, Aug 29, 2016 at 07:00:16PM -0700, Stephen Day wrote:
> >  type manifest struct {
> > -	Config descriptor   `json:"config"`
> > -	Layers []descriptor `json:"layers"`
> > +	Config *specs.Descriptor  `json:"config"`
> 
> No need for ptr here.

I expect this is another instance of [1].  If so, can we hash that
discussion out there?  If not, can you explain why you don't want a
pointer here and how that differs from why you don't want a pointer
argument for Engine.Put?

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76720093

stevvooe commented on image/refs/layout/tar.go:
```diff
@@ -0,0 +1,149 @@
+	"os"
+	"strings"
+
+	casLayout "github.com/opencontainers/image-spec/image/cas/layout"
```
avoid camelcase for package names.

wking replied:
On Mon, Aug 29, 2016 at 07:01:09PM -0700, Stephen Day wrote:
> > +	casLayout "github.com/opencontainers/image-spec/image/cas/layout"
> 
> avoid camelcase for package names.

I need to distinguish this from image/layout.  Do you have an
alternative local name you'd like me to use?

stevvooe replied:
`caslayout`.

stevvooe commented on image/refs/layout/tar.go:
```diff
@@ -0,0 +1,149 @@
+
+	casLayout "github.com/opencontainers/image-spec/image/cas/layout"
+	imageLayout "github.com/opencontainers/image-spec/image/layout"
+	utilLayout "github.com/opencontainers/image-spec/image/layout"
```
Two packages with different aliases?

stevvooe commented on cmd/oci-image-tool/cas_put.go:
```diff
@@ -0,0 +1,92 @@
+
+	digest += "\n"
+
+	n, err := io.WriteString(state.stdout, digest)
```
`fmt.Fprintln` will do this all in one go.

stevvooe commented on image/layout/tar.go:
```diff
@@ -0,0 +1,236 @@
+
+		if header.Name == name {
+			found = true
+			err = writeTarEntry(ctx, tarWriter, name, reader, size)
```
How does this avoid overwriting all subsequent entries in the tar file?

wking replied:
On Mon, Aug 29, 2016 at 07:08:33PM -0700, Stephen Day wrote:
> > +		if header.Name == name {
> > +			found = true
> > +			err = writeTarEntry(ctx, tarWriter, name, reader, size)
> 
> How does this avoid overwriting all subsequent entries in the tar file?

Because we're writing to a buffer [1] and only copy that buffer over
the file later [2].  More details on the implementation, the corners
it cuts, potential problems with it, and reasons why I went this way
anyway in the 3f19922d commit message.  And as I say there, folks
looking for a more reliable implementation should use a
filesystem-backed engine (which I haven't written up yet, but am happy
to add here if you like [3])

[1]: https://github.com/opencontainers/image-spec/pull/159/files#diff-745d1b74657ed8e87ac8c4e3f520f73aR77
[2]: https://github.com/opencontainers/image-spec/pull/159/files#diff-745d1b74657ed8e87ac8c4e3f520f73aR135
[3]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76726544

stevvooe replied:
I don't get the point of this model if it is only useful for a file-backed engine.

Is there an engine that supports a buffer and batch model that doesn't require memory to be allocated for all resources?

wking replied:
On Tue, Aug 30, 2016 at 01:44:20PM -0700, Stephen Day wrote:
> > +			err = writeTarEntry(ctx, tarWriter, name, reader, size)
> 
> I don't get the point of this model if it is only useful for a
> file-backed engine.

WriteTarEntryByName and writeTarEntry are helpers that are only useful
for the tar-backed engines (and we have CAS and ref engines backed by
tar files).  Zip-backed engines (which I'll add if/when this PR lands)
will probably need similar helpers.  Directory-backed engines (which
I'll add if/when this PR lands unless folks want them to be part of
this PR) will probably not need similar helpers.

> Is there an engine that supports a buffer and batch model that
> doesn't require memory to be allocated for all resources?

You can avoid buffering the tar file in memory if you write the new
tar file to disk and copy it over the original path, replacing the
file descriptor you were using with one that references the new file.
Any other engines which were using file descriptors referencing the
old instance will not notice the change.

I'm ok with doing that, but it means we'll probably want to adjust the
TarEngine functions to store the path (where they currently store the
ReadWriteSeekCloser) and have them open (and possibly
write/rename-over) the file for every operation.  That's doable, but
the current implementation is sufficient for testing, so I haven't
bothered.  If you'd rather me take that (or another) approach in this
PR, just let me know.

stevvooe commented:
What is the use case of these additions? Building up a tar file with random io is absurdly slow in practice and the implementation looks like it will overwrite following headers unless the blobs are added in order.

wking commented:
On Mon, Aug 29, 2016 at 07:09:37PM -0700, Stephen Day wrote:
> What is the use case of these additions?

The implementation in master assumes every backend will look like a
tar, and a tar walker is not a convenient/performant interface for
many types of engines (including directory-backed engines, HTTP-backed
engines, etc.).  This PR replaces the current tar-walker engine with
generic engines for CAS and refs, so you can build validators,
unpackers, publishers, etc. that don't care how the CAS blobs and refs
are stored as long as someone has written an Engine instance to wrap
the particular backend.

> Building up a tar file with random io is [snip: not good]

Agreed.  More on this in [2].

[1]: https://github.com/opencontainers/image-spec/pull/159#issue-160896428
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76726544

wking commented:
I pushed 1976fd6 → a139d21 which fixes some issues raised by
@stevvooe:

* Put writer -> reader in cas Engine.Put interface [1].
* imageLayout and utilLayout both pointed at image/layout [2].
* Replace io.WriteString with fmt.Fprintln for ‘cas put’ [3].

and rebases onto the current master.

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76719809
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76720273
[3]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76720480

stevvooe commented:
@wking This PR trades off a model that doesn't work for random access engines for one that doesn't work for tar. Can we find a model that can do both?

wking commented:
On Tue, Aug 30, 2016 at 01:45:53PM -0700, Stephen Day wrote:
> @wking This PR trades off a model that doesn't work for random
> access engines for one that doesn't work for tar. Can we find a
> model that can do both?

The current master has no write/delete support for tar.  This PR
*does* have solid support for reading from tar files (which is all
master has).  It cuts a few corners on writing to tar files, and we
can make other design choices if we want to put in the time to get
more robust tar-writing support [1].  But tar-writing support is never
going to be pretty for overwriting mutable refs, so I'm fine leaving
tar support as it is and pushing production users to saner backends if
you aren't comfortable with the PR's current tar-writing [1].

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76881527

wking commented:
I've pushed a139d21 → 6d71a76 dropping camelCase for package import
names [1].

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76720225

stevvooe commented:
@wking This PR attempts to apply a random access model to a tar file, such that each time you call the binary, a new tar file is created with the contents. If you try this with any non-trivial amount of data, it will be very slow. Furthermore, it buffers *all* of the data into memory to calculate the digest. Not only does it copy unnecessary data around, it also uses all the machines memory in doing so.

The ingest model for CAS is straightforward:

1. Ingest data to temporary, trusted location, while hashing.
2. Move atomically into final location.

Typically, these systems end up being CPU bound on hash speed while using very little memory for buffering.

These problems have been solved in other software systems, in Go, so it is not hard. In particular, the digest package from the docker/distribution package provided validators to due this in line. Much of the complexity in the registry is actually around coordinating temporary write locations. This is why I had asked you to take a peak at those other projects before proceeding too far down this path.

stevvooe commented on image/cas/layout/tar.go:
```diff
@@ -0,0 +1,108 @@
+
+// Put adds a new blob to the store.
+func (engine *TarEngine) Put(ctx context.Context, reader io.Reader) (digest string, err error) {
+	data, err := ioutil.ReadAll(reader)
```
Let's avoid allocating all the memory for this.

The idea would be to string the read of this content through the hash calculation, into a temp file.

wking replied:
On Tue, Aug 30, 2016 at 03:02:22PM -0700, Stephen Day wrote:
> > +// Put adds a new blob to the store.
> > +func (engine *TarEngine) Put(ctx context.Context, reader io.Reader) (digest string, err error) {
> > +	data, err := ioutil.ReadAll(reader)
> 
> Let's avoid allocating all the memory for this.
>
> The idea would be to string the read of this content through the
> hash calculation, into a temp file.

More on this in [1,2].  I don't think we need to care that much for
tar-writes.  I'm willing to implement it, but want to make sure you
think we do need to care that much.

[1]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243598892
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76890861

stevvooe commented on image/cas/layout/tar.go:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	size := int64(len(data))
+	hash := sha256.Sum256(data)
```
Use the writer version of this, along with `io.TeeReader`.

wking replied:
On Tue, Aug 30, 2016 at 03:02:59PM -0700, Stephen Day wrote:
> > +// Put adds a new blob to the store.
> > +func (engine *TarEngine) Put(ctx context.Context, reader io.Reader) (digest string, err error) {
> > +	data, err := ioutil.ReadAll(reader)
> > +	if err != nil {
> > +		return "", err
> > +	}
> > +
> > +	size := int64(len(data))
> > +	hash := sha256.Sum256(data)
> 
> Use the writer version of this, along with `io.TeeReader`.

The reason I don't, as discussed in cda16d26b, is that we aren't
buffering to a temporary disk location and we need to know the hash
before we can *start* writing to the tar file.  And again [1],
tar-write support as this PR stands is just good enough to pass most
unit tests.  I can reroll the tar backend to write to temporary files
if you like [2], but I personally don't see the need to make
tar-writing any more than a toy.  Anybody who has a lot of writing to
do should be using a directory-backed engine or similar which supports
atomic renames.

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76727482
[2]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243598892

wking commented:
On Tue, Aug 30, 2016 at 03:01:07PM -0700, Stephen Day wrote:
> @wking This PR attempts to apply a random access model to a tar
> file, such that each time you call the binary, a new tar file is
> created with the contents.

I don't see any way around that with per-call puts and a tar-backed
image-layout.  I'd strongly recommend folks who expect to do a lot of
writing write to a directory-backed image-layout, and then tar it up
when they're done.  Or call Engine.Put multiple times from a library
(which we could make reasonable performant in the new-key case if we
avoided writing the two zero blocks to the end of the tar file until
save time).

> Furthermore, it buffers *all* of the data into memory to calculate
> the digest.

It's copying the whole tar file via an in-memory buffer anyway, so
caching a single blob in memory isn't that big a deal.

> The ingest model for CAS is straightforward:
> 
> 1. Ingest data to temporary, trusted location, while hashing.
> 2. Move atomically into final location.

Which works beautifully for directory-backed engines.  Do you need me
to do the same for the tar-backed engine?  I will, but I'm personally
fine with the current tar-backend's terrible performance, because
folks will only ever use it for occasional tar creation and engine
testing.

wking commented:
I've pushed 6d71a76 → aabcdeb adjusting refs.Engine.List to use a
callback instead of channels [1].

The main outstanding issue from @stevvooe's recent review is whether I
need to reroll the tar handling to use the filesystem for scratch
space instead of using memory for scratch space [2 and many other
recent comments ;)].  I don't think we need to put in the work to make
tar writing more performant [3], but I'll put in the time if it's a
blocker for landing this PR.

[1]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76874690
[2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76881527
[3]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243587158

wking commented:
I've pushed aabcdeb → 4a6f6b7 rebasing onto master.  Changes:

* Resolved conflicts with #201 in favor of not changing the #159
  implementation (which already handles the missing-blob case since
  July [1,2]).

* Added man pages for the new commands to catch up with #180.

[1]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-230083383
[2]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-230175201

wking commented:
I've pushed 4a6f6b7 → 5f7a548 adding a comment to ListNameCallback
[1].

[1]: https://travis-ci.org/opencontainers/image-spec/builds/156439453#L324

wking commented:
I've pushed 5f7a548 → 9f4e036 fixing an unreachable return [1] which
snuck in with [2], and Travis is happy again :p.

[1]: https://travis-ci.org/opencontainers/image-spec/builds/156442741#L324
[2]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243668016

wking commented:
I've pushed c071139 → 7e96d03, rebasing around #212, #229, and #230
(master has been active in the past few days ;).  I'm still waiting on
maintainer feedback for:

On Tue, Aug 30, 2016 at 11:08:41PM -0700, W. Trevor King wrote [1]:
> The main outstanding issue from @stevvooe's recent review is whether I
> need to reroll the tar handling to use the filesystem for scratch
> space instead of using memory for scratch space [2 and many other
> recent comments ;)].  I don't think we need to put in the work to make
> tar writing more performant [3], but I'll put in the time if it's a
> blocker for landing this PR.
> 
> …
> [2]: https://github.com/opencontainers/image-spec/pull/159#discussion_r76881527
> [3]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243587158

I'm expecting something like one of:

a. “We agree that the current PR's tar handling isn't performant, but
   that it's not *broken* for constructing small image-layout tarballs
   from scratch, for append-only CAS operations, for new-key ref
   operations, or for ‘overwriting the same key with larger serialized
   descriptor JSON’ ref operations [2].  That's enough for the
   tar-backed engines, and we'll accept it for this PR.  We realize
   that there is a chance of a corrupted tarball if a writer crashes
   (discussed in 469bbb6e) and are ok with that because nobody should
   be using the tar-writer in production if they aren't prepared to
   catch crashed writes and start over from scratch.”,

b. “You need to use temporary files to avoid holding blobs in memory
   while you hash their contents and to avoid holding tarballs in
   memory while you rewrite them.  Using TempFile [3] with the default
   TempDir [4] is fine for this.  We realize that moving a blob from a
   temporary-disk file into a tarball is not an atomic operation, and
   that copying from TempDir over the original tar path may require
   futher copies [5].  We also realize that this will slow the
   read-only use-case, because readers will have to stat or re-open
   the file to detect changes made by other process [6].  We still
   want you to use temporary files.”, or

c. “Everything from (b), except you cannot rewrite the whole tar file
   when appending a new entry.  You must seek to before the trailing
   two zero-byte records and write into the existing tar file.  We
   realize that this leaves a window where the content of the tar file
   at the original tar path are incomplete, and accept the risk of
   corruption if a writer crashes (see (a)) in return for the
   efficiency gains from avoiding (b)'s temporary tar file.”.

[1]: https://github.com/opencontainers/image-spec/pull/159#issuecomment-243668016
[2]: Is there really no truncate(2) in Go's high-level file ops?  More
     in 469bbb6e.
[3]: https://golang.org/pkg/io/ioutil/#TempFile
[4]: https://golang.org/pkg/os/#TempDir
[5]: E.g. if TempFile and the original tar path are on different
     filesystems.
[6]: Both readers and writers currently seek back to the beginning of
     the file without re-opening it.

wking commented:
Moved to opencontainers/image-tools#5.
